python/cudf/cudf/core/mixins/mixin_factory.py

```diff
@@ -0,0 +1,218 @@
+       APIs usually accept an enum to delineate the specific operation to
+       perform, e.g. binary operations use the `binary_operator` enum when
+       calling the `binary_operation` function. cuDF Python mimics this
+       structure by having operations within a category delegate to a common
+       internal function (e.g. DataFrame.__add__ calls DataFrame._binaryop).
+    2. Many cuDF classes implement similar operations (e.g. `sum`) via
+       delegation to lower-level APIs before reaching a libcudf C++ function
+       call. As a result, many API function calls actually involve multiple
+       delegations to lower-level APIs that can look essentially identical. An
+       example of such a sequence would be DataFrame.sum -> DataFrame._reduce
+       -> Column.sum -> Column._reduce -> libcudf.
+
+    This factory creates mixins for a category of operations implemented by via
+    this delegator pattern. The resulting mixins make it easy to share common
+    functions across various classes while also providing a common entrypoint
+    for implementing the centralized logic for a given category of operations.
+    Its usage is best demonstrated by example below.
+
+    Parameters
+    ----------
+    mixin_name : str
+        The name of the class. This argument should be the same as the object
+        that this function's output is assigned to, e.g.
+        :code:`Baz = _create_delegating_mixin("Baz", ...)`.
+    docstring : str
+        The documentation string for the mixin class.
+    category_name : str
+        The category of operations for which a mixin is being created. This
+        name will be used to define or access the following attributes as shown
+        in the example below:
+            - f'_{category_name}_DOCSTRINGS'
+            - f'_VALID_{category_name}S'
+            - f'_SUPPORTED_{category_name}S'
+    category_operation_name : str
+        The name given to the core function implementing this category of
+        operations.  The corresponding function is the entrypoint for child
+        classes.
+    supported_ops : List[str]
+        The list of valid operations that subclasses of the resulting mixin may
+        request to be implemented.
+
+    Examples
+    --------
+    >>> MyFoo = _create_delegating_mixin(
+    ...     "MyFoo", "MyFoo's docstring", "FOO", "do_foo", {"foo1", "foo2"}
+    ... )
+
+    >>> # The above is equivalent to defining a class like so:
+    ... class MyFoo:
+    ...     # The set of valid foo operations.
+    ...     _VALID_FOOS = {"foo1", "foo2"}
+    ...
+    ...     # This is the method for child classes to override. Note that the
+    ...     # first parameter is always called "op".
+    ...     def do_foo(self, op, *args, **kwargs):
+    ...         raise NotImplementedError
+
+    >>> # MyFoo can be used as follows.
+    >>> class BarImplementsFoo(MyFoo):
+    ...     _SUPPORTED_FOOS = ("foo1",)
```

Comment:
```suggestion
    ...     _VALID_FOOS = ("foo1",)
```

Reply:
It's probably better to enforce `_VALID_FOOS` as an ordered iterable? Probably wouldn't matter but I don't feel quite easy when you are programmatically generating methods in an undefined order...

Reply:
Again, good catch on me swapping VALID and SUPPORTED in the example.

Regarding ordering, I don't think this is an issue. We can fix it later if for some reason we find an issue, but as of now even if `_VALID_FOOS` is an ordered iterable that order will be destroyed when we walk the MRO to construct the full set of all inherited methods. However, in light of some of the other comments, I'm considering another change that might address this along with fixing some other issues. I'll report back in a bit.

Reply:
Actually it turns out that the change to resolve the inheritance issue makes it even more important that we use a set to collect all ops, whereas I was hoping that we could avoid performing that collection from parent classes altogether. While we _could_ potentially enforce an ordering by requiring an ordered iterable and using a dictionary instead of a set, I don't think we would benefit much from it, especially given that with inherited reductions we would end up with a relatively arbitrary albeit deterministic order of definitions. Let's not worry about this for now; we can revisit if we notice this becoming a real issue.